File: LeetCode-2023/1658.将-x-减到-0-的最小操作数.py
#
# @lc app=leetcode.cn id=1658 lang=python3
#
# [1658] 将 x 减到 0 的最小操作数
#

# @lc code=start
class Solution:
    def minOperations(self, nums: List[int], x: int) -> int:
        if x < 1:
            return -1
        sum = 0
        for i in nums:
            sum = sum + i
        target = sum - x

        left, right = 0, 0
        now = 0
        result = -1
        while right < len(nums) :
            now = now + nums[right]
            if now == target:
                result = max(result, right - left + 1)
            while left <= right and now >= target:
                now = now - nums[left]
                left = left + 1
                if now == target:
                    result = max(result, right - left + 1)
            right = right + 1
        if result > -1:
            return len(nums) - result
        else:
            return -1
# @lc code=end

File: LeetCode-2023/1658.将-x-减到-0-的最小操作数.java
/*
 * @lc app=leetcode.cn id=1658 lang=java
 *
 * [1658] 将 x 减到 0 的最小操作数
 */

// @lc code=start
class Solution {
    public int minOperations(int[] nums, int x) {
        if(x < 1) return -1;
        int sum = 0;
        for(int i = 0; i < nums.length; i++) {
            sum += nums[i];
        }
        int target = sum - x;

        int left = 0, right = 0;
        int now = 0;
        int result = -1;
        while(right < nums.length) {
            now += nums[right];
            if(now == target) {
                result = Math.max(result, right - left + 1);
            }
            while(now >= target && left <= right) {
                now -= nums[left];
                left++;
                if(now == target) {
                    result = Math.max(result, right - left + 1);
                }
            }
            right++;
        }
        return result > -1 ? nums.length - result : -1;
    }
}
// @lc code=end

